[prefer-nullish-coalescing] eslint --fix outputs broken code

<!--
If you have a problem with a specific rule, please begin your issue title with [rulename] to make it easier to search for.
I.e. "[no-unused-vars] False positive when fooing the bar"

Please don't ignore this template.

If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
Are you opening an issue because the rule you're trying to use is not found?
🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨
1) Check the releases log: https://github.com/typescript-eslint/typescript-eslint/releases
    -  If the rule isn't listed there, then chances are it hasn't been released to the main npm tag yet.
2) Try installing the `canary` tag: `npm i @typescript-eslint/eslint-plugin@canary`.
    - The canary tag is built for every commit to master, so it contains the bleeding edge build.
3) If ESLint still can't find the rule, then consider reporting an issue.
-->

**Repro**

<!--
Include a ***minimal*** reproduction case.
The more irrelevant code/config you give, the harder it is for us to investigate.
-->

`.eslintrc.json`
```JSON
{
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "parserOptions": {
    "ecmaVersion": 2018,
    "sourceType": "module",
    "project": "./tsconfig.json"
  },
  "rules": {
    "@typescript-eslint/prefer-nullish-coalescing": "error"
  }
}
```

`tsconfig.json`
```JSON
{
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "lib",
    "strict": true
  },
  "include": [
    "src/**/*"
  ]
}
```

`src/index.ts`
```TS
declare const a: string | undefined;
declare const b: string;
declare const c: string;

const x = a || b || c;
```

**Expected Result**
The last line should be fixed like this:
```TS
const x = (a ?? b) || c;
```
Here parentheses are required because nullish coalescing operators and logical operators cannot be mixed.
Since logical operator `||` is left-associative, it is natural to add parentheses in left-to-right manner.

**Actual Result**
`eslint --fix` outputs this:
```TS
const x = a ?? b || c;
```

This is not a valid syntax and we get the following error.

```
$ /path/to/project/node_modules/.bin/tsc --build .
src/index.ts:5:16 - error TS5076: '||' and '??' operations cannot be mixed without parentheses.

5 const x = a ?? b || c;
                 ~~~~~~


Found 1 error.
```

**Additional Info**

<!--
Did eslint throw an exception?

Please run your lint again with the --debug flag, and dump the output below.
i.e. eslint --ext ".ts,.js" src --debug
-->

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.9.0` |
| `@typescript-eslint/parser`        | `2.9.0` |
| `TypeScript`                       | `3.7.2` |
| `ESLint`                           | `6.7.2` |
| `node`                             | `12.13.0` |
| `npm`                              | `6.13.1` |

--- a/packages/eslint-plugin/src/rules/prefer-nullish-coalescing.ts
+++ b/packages/eslint-plugin/src/rules/prefer-nullish-coalescing.ts
@@ -103,6 +103,24 @@ export default util.createRule<Options, MessageIds>({
           util.NullThrowsReasons.MissingToken('operator', node.type),
         );
 
+        function* fix(
+          fixer: TSESLint.RuleFixer,
+        ): IterableIterator<TSESLint.RuleFix> {
+          if (node.parent && util.isLogicalOrOperator(node.parent)) {
+            // '&&' and '??' operations cannot be mixed without parentheses (e.g. a && b ?? c)
+            if (
+              node.left.type === AST_NODE_TYPES.LogicalExpression &&
+              !util.isLogicalOrOperator(node.left.left)
+            ) {
+              yield fixer.insertTextBefore(node.left.right, '(');
+            } else {
+              yield fixer.insertTextBefore(node.left, '(');
+            }
+            yield fixer.insertTextAfter(node.right, ')');
+          }
+          yield fixer.replaceText(barBarOperator, '??');
+        }
+
         const fixer =
           isMixedLogical || forceSuggestionFixer
             ? // suggestion instead for cases where we aren't sure if the fixer is completely safe
@@ -110,17 +128,11 @@ export default util.createRule<Options, MessageIds>({
                 suggest: [
                   {
                     messageId: 'preferNullish',
-                    fix(fixer: TSESLint.RuleFixer): TSESLint.RuleFix {
-                      return fixer.replaceText(barBarOperator, '??');
-                    },
+                    fix,
                   },
                 ],
               } as const)
-            : {
-                fix(fixer: TSESLint.RuleFixer): TSESLint.RuleFix {
-                  return fixer.replaceText(barBarOperator, '??');
-                },
-              };
+            : { fix };
 
         context.report({
           node: barBarOperator,

--- a/packages/eslint-plugin/src/util/astUtils.ts
+++ b/packages/eslint-plugin/src/util/astUtils.ts
@@ -1,7 +1,7 @@
 import {
-  TSESTree,
-  AST_TOKEN_TYPES,
   AST_NODE_TYPES,
+  AST_TOKEN_TYPES,
+  TSESTree,
 } from '@typescript-eslint/experimental-utils';
 
 const LINEBREAK_MATCHER = /\r\n|[\r\n\u2028\u2029]/;
@@ -43,6 +43,15 @@ function isOptionalOptionalChain(
 }
 
 /**
+ * Returns true if and only if the node represents logical OR
+ */
+function isLogicalOrOperator(node: TSESTree.Node): boolean {
+  return (
+    node.type === AST_NODE_TYPES.LogicalExpression && node.operator === '||'
+  );
+}
+
+/**
  * Determines whether two adjacent tokens are on the same line
  */
 function isTokenOnSameLine(
@@ -59,5 +68,6 @@ export {
   isOptionalChainPunctuator,
   isOptionalOptionalChain,
   isTokenOnSameLine,
+  isLogicalOrOperator,
   LINEBREAK_MATCHER,
 };

--- a/packages/eslint-plugin/tests/rules/prefer-nullish-coalescing.test.ts
+++ b/packages/eslint-plugin/tests/rules/prefer-nullish-coalescing.test.ts
@@ -317,7 +317,7 @@ declare const a: ${type} | ${nullish};
 declare const b: ${type} | ${nullish};
 declare const c: ${type} | ${nullish};
 declare const d: ${type} | ${nullish};
-a ?? b || c && d;
+(a ?? b) || c && d;
               `.trimRight(),
             },
           ],
@@ -367,7 +367,7 @@ declare const a: ${type} | ${nullish};
 declare const b: ${type} | ${nullish};
 declare const c: ${type} | ${nullish};
 declare const d: ${type} | ${nullish};
-a && b ?? c || d;
+a && (b ?? c) || d;
               `.trimRight(),
             },
           ],
@@ -463,5 +463,30 @@ x ?? 'foo';
         },
       ],
     },
+
+    // https://github.com/typescript-eslint/typescript-eslint/issues/1290
+    ...nullishTypeInvalidTest((nullish, type) => ({
+      code: `
+declare const a: ${type} | ${nullish};
+declare const b: ${type};
+declare const c: ${type};
+a || b || c;
+      `.trimRight(),
+      output: `
+declare const a: ${type} | ${nullish};
+declare const b: ${type};
+declare const c: ${type};
+(a ?? b) || c;
+      `.trimRight(),
+      errors: [
+        {
+          messageId: 'preferNullish',
+          line: 5,
+          column: 3,
+          endLine: 5,
+          endColumn: 5,
+        },
+      ],
+    })),
   ],
 });

